### PR TITLE
HTTP/2 Headers Type Updates

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The HTTP/2 RFC (https://tools.ietf.org/html/rfc7540#section-8.1.2) indicates that header names consist of ASCII characters. We currently use ByteString to represent HTTP/2 header names. The HTTP/2 RFC (https://tools.ietf.org/html/rfc7540#section-10.3) also eludes to header values inheriting the same validity characteristics as HTTP/1.x. Using AsciiString for the value type of HTTP/2 headers would allow for re-use of predefined HTTP/1.x values, and make comparisons more intuitive. The Headers<T> interface could also be expanded to allow for easier use of header types which do not have the same Key and Value type.

Motivation:
- Change Headers<T> to Headers<K, V>
- Change Http2Headers<ByteString> to Http2Headers<CharSequence, CharSequence>
- Remove ByteString. Having AsciiString extend ByteString complicates equality comparisons when the hash code algorithm is no longer shared.

Result:
Http2Header types are more representative of the HTTP/2 RFC, and relationship between HTTP/2 header name/values more directly relates to HTTP/1.x header names/values.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @normanmaurer Spikhalskiy  - FYI

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java [hunk 1]
```diff
@@ -333,12 +333,36 @@ private static void validateChar(char character) {
...
-            if (character > 127) {
+            if (value < 0) {
                 throw new IllegalArgumentException("a header name cannot contain non-ASCII character: " +
-                        character);
+                        value);
+            }
+        }
+    }
+
+    private static void validateHeaderNameElement(char value) {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java]: @normanmaurer - I broke this out for testing to avoid having to `1. cast to int. 2. mask 0xFF 3. cast to char` for each `byte` if AsciiString is used.  I haven't yet verified if this is justified yet...

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java [hunk 2]
```diff
@@ -86,7 +87,7 @@
      */
     public static final HttpMethod CONNECT = new HttpMethod("CONNECT");
 
-    private static final Map<String, HttpMethod> methodMap = new HashMap<String, HttpMethod>();
+    private static final Map<String, HttpMethod> methodMap = new TreeMap<String, HttpMethod>();
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: this was on the hot path during benchmarks ... I may pull this out into another PR ... or at the very least merge it back to 4.0.

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java [hunk 3]
```diff
@@ -352,13 +376,9 @@ private static void validateChar(char character) {
 
         @Override
         public CharSequence convertObject(Object value) {
-            checkNotNull(value, "value");
             if (value instanceof CharSequence) {
                 return (CharSequence) value;
             }
-            if (value instanceof Number) {
```
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java]: why remove this?

File: codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java [hunk 4]
```diff
@@ -25,7 +25,7 @@
 import io.netty.handler.codec.EmptyHeaders;
 import io.netty.handler.codec.Headers;
 
-public class EmptyHttpHeaders extends EmptyHeaders<CharSequence> implements HttpHeaders {
+public class EmptyHttpHeaders extends EmptyHeaders<CharSequence, CharSequence> implements HttpHeaders {
```
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java]: is anything changed in this file?

Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/EmptyHttpHeaders.java]: oh nm ... it is changed.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java [hunk 5]
```diff
@@ -44,23 +45,15 @@ public static void encoderHeader(CharSequence name, CharSequence value, ByteBuf
...
         }
     }
 
-    private static void writeAsciiString(ByteBuf buf, int offset, AsciiString value, int valueLen) {
-        ByteBufUtil.copy(value, 0, buf, offset, valueLen);
-    }
-
     private static void writeCharSequence(ByteBuf buf, int offset, CharSequence value, int valueLen) {
-        for (int i = 0; i < valueLen; i ++) {
+        for (int i = 0; i < valueLen; ++i) {
```
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java]: FYI, this will only work for US-ASCII.  Is the assumption that this is ok since HTTP/1.x headers only support US-ASCII?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 2 above)
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: Didn't @buchgr make this a `TreeMap` at one point?

Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: I'm fine with it ... do you have benchmarks?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+package io.netty.handler.codec.http2;
+
+import io.netty.handler.codec.DefaultHeaders;
+import io.netty.handler.codec.UnsupportedValueConverter;
+import io.netty.handler.codec.ValueConverter;
+
+import static io.netty.util.AsciiString.CASE_INSENSITIVE_HASHER;
+import static io.netty.util.AsciiString.CASE_SENSITIVE_HASHER;
+
+final class CharSequenceMap<V> extends DefaultHeaders<CharSequence, V> {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: Is this class needed?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java [hunk 7]
```diff
@@ -69,203 +83,207 @@ public DefaultHttp2Headers() {
...
     @Override
-    public Http2Headers add(ByteString name, Iterable<? extends ByteString> values) {
+    public Http2Headers add(CharSequence name, Iterable<? extends CharSequence> values) {
         super.add(name, values);
         return this;
     }
 
     @Override
-    public Http2Headers add(ByteString name, ByteString... values) {
+    public Http2Headers add(CharSequence name, CharSequence... values) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: Could we make the base class automatically return the correct type with generics?  It's unfortunate to have to implement every method just for altering the return type.  Alternatively, we could get rid of this class entirely and just have a factory class:

```java
public final class Http2CodecUtil {
...
    public static HttpHeaders newHeaders() {
        // Create HttpHeaders and configure it the way we want.
    }
...
}
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java [hunk 8]
```diff
@@ -72,9 +72,9 @@ public Http2ChannelClosedException() {
         private static final long serialVersionUID = 1326785622777291198L;
         private final int lastStreamId;
         private final long errorCode;
-        private final ByteString debugData;
+        private final AsciiString debugData;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Hmm ... not sure if this is right.  Debug data isn't necessarily text.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Maybe that's fine? It's just a slightly odd use of `AsciiString` in that case?

File: codec/src/main/java/io/netty/handler/codec/Headers.java [hunk 9]
```diff
@@ -19,7 +19,14 @@
...
 
-public interface Headers<T> extends Iterable<Entry<T, T>> {
+/**
+ * Common interface for {@link Headers} which represents a mapping of key to value.
+ * Duplicate keys may be allowed by implementations.
+ *
+ * @param <K> the type of the header name.
+ * @param <V> the type of the header value.
+ */
+public interface Headers<K, V> extends Iterable<Entry<K, V>> {
```
Comment (nmittler) [inline codec/src/main/java/io/netty/handler/codec/Headers.java]: Do we currently have a use case for anything other than `Headers<CharSequence, CharSequence>`? If not, perhaps we should drop one or both parameters altogether?

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 10]
```diff
@@ -35,135 +39,396 @@
...
-        public boolean equals(CharSequence a, CharSequence b) {
-            return AsciiString.contentEquals(a, b);
-        }
-    };
+    private static final char MAX_CHAR_VALUE = 255;
+    private static final int HASH_CODE_PRIME = 31;
 
     /**
-     * Factory which uses the {@link #AsciiString(byte[], int, int, boolean)} constructor.
+     * If this value is modified outside the constructor then call {@link #arrayChanged()}.
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/AsciiString.java]: Maybe this should be mentioned in the class-level javadoc?

File: common/src/main/java/io/netty/util/internal/MathUtil.java [hunk 11]
```diff
@@ -35,4 +35,16 @@ public static int findNextPositivePowerOfTwo(final int value) {
...
+
+    /**
+     * Determine if the requested {@code index} and {@code length} will fit within {@code capacity}.
+     * @param index The starting index.
+     * @param length The length which will be utilized (starting from {@code index}).
+     * @param capacity The capacity that {@code index + length} is allowed to be within.
+     * @return {@code true} if the requested {@code index} and {@code length} will fit within {@code capacity}.
+     * {@code false} if this would result in an index out of bounds exception.
+     */
+    public static boolean isOutOfBounds(int index, int length, int capacity) {
```
Comment (nmittler) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: +1 I thought this should have been a utility :)

**Comment by nmittler** [comment]
@Scottmitch took a first pass ... great job!

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java]: Just pruning code a bit. `toString()` is already done at the end of the block.  I can revert if you like (if you think the slightly faster path for `Number` at the expense of other types is justified).

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java]: Yes. https://tools.ietf.org/html/rfc7230#section-3.2. where `VCHAR (any visible [USASCII] character)`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java (hunk 6 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: This is a bit of a hack to leverage the DefaultHeader's ability to take a `HashingStrategy`. We had some `Map` objects internally that depended on being able to compare `CharSequence` objects (mostly `String` and `AsciiString`) using `contentEquals*` instead of `equals`.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java]: I think I see what you are saying ... because this is not in the HTTP module it may be a bit misleading?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java (hunk 7 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java]: Good idea. Let me try to make the generics take care of the return type.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 8 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: We are back to the gray area of needing an immutable byte string ... is this even used? If we still need it we have a few options:

1. Leave as is ... comment it should be treated as a byte array.
2. Change to `byte[]`
3. Change to `ByteBuf` ... we would need to be careful about reference count in this case.

WDYT?

File: codec/src/main/java/io/netty/handler/codec/Headers.java (hunk 9 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/Headers.java]: All types are now `Headers<CharSequence, CharSequence>`. However I'm not sure if it is necessary or if there would be significant gains to limit the generic nature of this class. If anything could we handle this in a followup PR?

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 10 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Done.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: I have a few flight recorder results...however I think it warrants a closer look. Let me pull out into a another PR so we can look at this independently

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpMethod.java]: See https://github.com/netty/netty/pull/4424

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 11 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: :smile:

thanks for pointing this out ... need to move this around in 4.0 too

**Comment by Scottmitch** [comment]
@nmittler - I addressed most of your comments. Some followup questions on others. PTAL.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 8 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: I'm tempted to just use byte[] .. WDYT?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java [hunk 12]
```diff
@@ -89,9 +89,9 @@ public long errorCode() {
             return errorCode;
         }
 
-        public ByteString debugData() {
-            return debugData;
-        }
+//        public AsciiString debugData() {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Is this intended to be commented out?

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java (hunk 3 above)
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java]: Nope, that SGTM ... thanks!

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java (hunk 5 above)
Comment (nmittler) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java]: No, it's fine ... It clearly states `HttpHeaders` in the name ... that's sufficient I think.

**Comment by nmittler** [comment]
@Scottmitch changes look great ... just a couple minor comments.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpHeadersEncoder.java]: Ok good... my previous statement was also false ... it is in the HTTP module :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 8 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: `byte[]` works for me.

We can impose a limit on the amount we copy with a followup PR if necessary.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 12 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: No it wasn't

**Comment by Scottmitch** [comment]
@normanmaurer - Did you want to take a look before I pull in?

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java [hunk 13]
```diff
@@ -33,6 +33,7 @@
 import java.nio.channels.ScatteringByteChannel;
 import java.nio.charset.Charset;
 
+import static io.netty.util.internal.MathUtil.isOutOfBounds;
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: not sure this really belongs to MathUtils... Maybe ByteBufUtil ?

File: buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java [hunk 14]
```diff
@@ -328,7 +328,7 @@ static void getBytes(AbstractByteBuf buf, long addr, int index, ByteBuffer dst)
     static void setBytes(AbstractByteBuf buf, long addr, int index, ByteBuf src, int srcIndex, int length) {
         buf.checkIndex(index, length);
         checkNotNull(src, "src");
-        if (AbstractByteBuf.isInvalid(srcIndex, length, src.capacity())) {
+        if (isOutOfBounds(srcIndex, length, src.capacity())) {
             throw new IndexOutOfBoundsException("srcIndex: " + srcIndex);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java]: Also I would prefer if you revert all of these changes above as these not belong to the header change if you ask me.

File: codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java (hunk 1 above)
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/DefaultHttpHeaders.java]: ok

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+import io.netty.handler.codec.UnsupportedValueConverter;
+import io.netty.handler.codec.ValueConverter;
+
+import static io.netty.util.AsciiString.CASE_INSENSITIVE_HASHER;
+import static io.netty.util.AsciiString.CASE_SENSITIVE_HASHER;
+
+final class CharSequenceMap<V> extends DefaultHeaders<CharSequence, V, CharSequenceMap<V>> {
+    public CharSequenceMap() {
+        this(true);
+    }
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: add empty line

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    public CharSequenceMap(boolean caseSensative) {
+        this(caseSensative, UnsupportedValueConverter.<V>instance());
+    }
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: add empty line

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java [hunk 17]
```diff
@@ -89,7 +88,7 @@ public long errorCode() {
             return errorCode;
         }
 
-        public ByteString debugData() {
+        public byte[] debugData() {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Do we need to call `close()` on the array ? OR should me maybe return a ByteBuf or ByteBuffer ?

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+public final class UnsupportedValueConverter<V> implements ValueConverter<V> {
+    @SuppressWarnings("rawtypes")
+    private static final UnsupportedValueConverter INSTNACE = new UnsupportedValueConverter();
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: typo... INSTNACE -> INSTANCE

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    private UnsupportedValueConverter() { }
+
+    @SuppressWarnings("unchecked")
+    public static <V> UnsupportedValueConverter<V> instance() {
+        return (UnsupportedValueConverter<V>) INSTNACE;
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: see above

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+public final class UnsupportedValueConverter<V> implements ValueConverter<V> {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: javadocs

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 21]
```diff
@@ -20,150 +20,420 @@
...
+     *
+     * @return {@code -1} if the processor iterated to or beyond the end of the specified area.
+     *         The last-visited index If the {@link ByteProcessor#process(byte)} returned {@code false}.
+     */
+    public int forEachByte(int index, int length, ByteProcessor visitor) throws Exception {
+        if (isOutOfBounds(index, length, length())) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= index(" + index + ") <= start + length(" + length
+                    + ") <= " + "length(" + length() + ')');
+        }
+
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: kill empty line

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 22]
```diff
@@ -20,150 +20,420 @@
...
+     *
+     * @return {@code -1} if the processor iterated to or beyond the beginning of the specified area.
+     *         The last-visited index If the {@link ByteProcessor#process(byte)} returned {@code false}.
+     */
+    public int forEachByteDesc(int index, int length, ByteProcessor visitor) throws Exception {
+        if (isOutOfBounds(index, length, length())) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= index(" + index + ") <= start + length(" + length
+                    + ") <= " + "length(" + length() + ')');
+        }
+
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: kill empty line

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 23]
```diff
@@ -20,150 +20,420 @@
...
+    public byte byteAt(int index) {
+        // We must do a range check here to enforce the access does not go outside our sub region of the array.
+        // We rely on the array access itself to pick up the array out of bounds conditions
+        if (index < 0 || index >= length) {
+            throw new IndexOutOfBoundsException("index: " + index + " must be in the range [0," + length + ")");
+        }
+        // Try to use unsafe to avoid double checking the index bounds
+        if (PlatformDependent.hasUnsafe()) {
+            return PlatformDependent.getByte(value, index + offset);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: yay :)

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 24]
```diff
@@ -20,150 +20,420 @@
...
+
+    /**
+     * Determines if this {@code String} contains the sequence of characters in the {@code CharSequence} passed.
+     *
+     * @param cs the character sequence to search for.
+     * @return {@code true} if the sequence of characters are contained in this string, otherwise {@code false}.
+     */
+    public boolean contains(CharSequence cs) {
+        if (cs == null) {
+            throw new NullPointerException();
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: Use ObjectUtil.requireNonNull ?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: also if indexOf(...) already check for null you may just remove this check here.

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java (hunk 13 above)
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: I'm not sure about MathUtils either, but it needs to go in the `common` module because it is used by `AsciiString` (which lives in the common module).  Maybe I should create an `ArrayUtils` in `common`?

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 11 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: Actually I'm not sure it really belongs to MathUtil... But not strong here

File: buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java (hunk 14 above)
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java]: The reason why it was done here is because `AsciiString` now uses it. I can pull it out, but I would prefer not to (just creates extra work). As I said in https://github.com/netty/netty/pull/4413#discussion_r43532383 I will port the changes to 4.0 too.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java (hunk 15 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java (hunk 16 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CharSequenceMap.java]: done

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java (hunk 13 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: Well it is not really related only to arrays as well as we use it in our direct buffers too :) So maybe it's just fine as MathUtil is an internal class anyway and we can move it later.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 17 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: > Do we need to call `close()` on the array ?

Do you mean `clone`?  I don't think that is necessary as it is already a copy from the original data.

> should me maybe return a ByteBuf or ByteBuffer ?

@nmittler and I discussed this above https://github.com/netty/netty/pull/4413#discussion_r43528906.  The issue with `ByteBuf` is the reference counting. The exception is used to complete a promise, which may not be expecting to have to release any `ByteBuf` payload in the `Exception`.  I guess we could use `nioBufferCount` and `nioBuffer` to see if we can get a `ByteBuffer` ... but if it doesn't we still may have to copy anyways.  Also it is a bit awkward exposing a `ByteBuffer` to the user. @nmittler - WDYT?

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java (hunk 18 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: done

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java (hunk 19 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: done

File: codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java (hunk 20 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/UnsupportedValueConverter.java]: done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 17 above)
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Yes `clone()` ... So maybe just return the array is fine as long.

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 24 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Yah I'll just remove it.

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 11 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: I'm not strong either ... I'm open for suggestions (it just has to live in `common` somewhere).

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 23 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: +1 for unsafe!

File: common/src/main/java/io/netty/util/internal/MathUtil.java (hunk 11 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: Let's just move it... we can move it again later if we want to as MathUtil is internal anyway

File: buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java (hunk 14 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java]: yeah I missed this... so just leave it

**Comment by Scottmitch** [comment]
@nmittler @normanmaurer - Want to take another look or are you satisfied?

**Comment by normanmaurer** [comment]
@Scottmitch I'm good...

File: codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java (hunk 17 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java]: Yeah, we expect ~1 GOAWAY per connection .... not worried about copying :)

**Comment by nmittler** [comment]
@Scottmitch LGTM

**Comment by Scottmitch** [comment]
Cherry-picked 4.0 (just some - d353d59) 4.1 (19658e9) master (1c93e04)

**Comment by Scottmitch** [comment]
Thanks for review @nmittler and @normanmaurer !
